go/vt/vtgate/planbuilder/operators/info_schema_planning.go

```diff
@@ -0,0 +1,314 @@
+limitations under the License.
+*/
+
+package operators
+
+import (
+	"strings"
+
+	"vitess.io/vitess/go/vt/vtgate/planbuilder/operators/ops"
+	"vitess.io/vitess/go/vt/vtgate/vindexes"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	"vitess.io/vitess/go/mysql/collations"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/servenv"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vtgate/engine"
+	"vitess.io/vitess/go/vt/vtgate/evalengine"
+	"vitess.io/vitess/go/vt/vtgate/planbuilder/plancontext"
+)
+
+// InfoSchemaRouting used for information_schema queries.
+// They are special because we usually don't know at plan-time
+// what keyspace the query go to, because we don't see normalized literal values
+type InfoSchemaRouting struct {
+	SysTableTableSchema []sqlparser.Expr
+	SysTableTableName   map[string]sqlparser.Expr
+	Table               *QueryTable
+}
+
+func (isr *InfoSchemaRouting) UpdateRoutingParams(_ *plancontext.PlanningContext, rp *engine.RoutingParameters) error {
+	rp.SysTableTableSchema = nil
+	for _, expr := range isr.SysTableTableSchema {
+		eexpr, err := evalengine.Translate(expr, &notImplementedSchemaInfoConverter{})
+		if err != nil {
+			return err
+		}
+		rp.SysTableTableSchema = append(rp.SysTableTableSchema, eexpr)
+	}
+
+	rp.SysTableTableName = make(map[string]evalengine.Expr, len(isr.SysTableTableName))
+	for k, expr := range isr.SysTableTableName {
+		eexpr, err := evalengine.Translate(expr, &notImplementedSchemaInfoConverter{})
+		if err != nil {
+			return err
+		}
+
+		rp.SysTableTableName[k] = eexpr
+	}
+	return nil
+}
+
+func (isr *InfoSchemaRouting) Clone() Routing {
+	return &InfoSchemaRouting{
+		SysTableTableSchema: slices.Clone(isr.SysTableTableSchema),
+		SysTableTableName:   maps.Clone(isr.SysTableTableName),
+		Table:               isr.Table,
+	}
+}
+
+func (isr *InfoSchemaRouting) updateRoutingLogic(ctx *plancontext.PlanningContext, expr sqlparser.Expr) (Routing, error) {
+	isTableSchema, bvName, out := extractInfoSchemaRoutingPredicate(expr, ctx.ReservedVars)
+	if out == nil {
+		return isr, nil
+	}
+
+	if isr.SysTableTableName == nil {
+		isr.SysTableTableName = map[string]sqlparser.Expr{}
+	}
+
+	if isTableSchema {
+		for _, s := range isr.SysTableTableSchema {
+			if sqlparser.Equals.Expr(out, s) {
+				// we already have this expression in the list
+				// stating it again does not add value
+				return isr, nil
+			}
+		}
+		isr.SysTableTableSchema = append(isr.SysTableTableSchema, out)
+	} else {
+		isr.SysTableTableName[bvName] = out
+	}
+	return isr, nil
+}
+
+func (isr *InfoSchemaRouting) Cost() int {
+	return 0
+}
+
+func (isr *InfoSchemaRouting) OpCode() engine.Opcode {
+	return engine.DBA
+}
+
+func (isr *InfoSchemaRouting) Keyspace() *vindexes.Keyspace {
+	// TODO: for some info schema queries, we do know which keyspace it will go to
+	// if we had this information, more routes could be merged.
+	return nil
+}
```

Comment:
what we can do is, if there is only 1 keyspace then we know for sure where the query will go to. This can help merge some routes.

Reply:
yep. follow up PR though, right?

Reply:
Having a task in information schema work for memory will work.